Cookbook/Technical-Documents/Architecture.md

```diff
@@ -0,0 +1,309 @@
+
+[https://github.com/Babylonpartners/Bento/tree/635d6d25e9152ba38aba8620e307431a0b8db4dd/BentoKit/BentoKit/Screen](https://github.com/Babylonpartners/Bento/tree/635d6d25e9152ba38aba8620e307431a0b8db4dd/BentoKit/BentoKit/Screen)
+
+```swift
+public protocol BoxViewModel: AnyObject {
+    associatedtype State
+    associatedtype Action
+
+    // `BoxViewModel` to `BoxRenderer` event flow
+    var state: Property<State> { get }
+
+    // `BoxRenderer` to `BoxViewModel` event flow (user action from UI)
+    func send(action: Action)
+}
+
+public protocol BoxRenderer {
+    typealias Sink<Action> = (Action) -> Void
+    associatedtype State
+    associatedtype Action
+    associatedtype SectionID: Hashable
+    associatedtype ItemID: Hashable
+
+    // `observer` as `BoxRenderer` to `BoxViewModel` event flow (user action from UI)
+    init(observer: @escaping Sink<Action>, ....)
+
+    func render(state: State) -> Screen<SectionID, ItemID>
+}
+```
+
+#### Unidirectional event flow
+
+1. `BoxRenderer.observer` is triggered with a new event
+2. `BoxViewModel.send(action:)` is invoked and changes state
+3. `state: Property<State>` emits a new state
+4. `Bento.BoxViewController` receives a new state, creating virtual view via `BoxRenderer.render(state:)` and apply diffing to real `UITableView`
+5. `UITableView` gets reloaded
+
+### `ViewModel`
+
+The minimal `ViewModel` example looks as follows:
+
+```swift
+class ViewModel: Bento.BoxViewModel {
+    let state: Property<State>
+    let routes: Signal<Route, NoError>
+
+    private let eventPipe = Feedback<Event, NoError>.input()
+
+    init(...) {
+        self.state = Property( // Defined in ReactiveFeedback
+            initial: .initial,
+            reduce: ViewModel.reduce,
+            feedbacks: [
+                actionsPipe.feedback,
+                ViewModel.whenSubmitting(...)
+            ]
+        )
+
+        self.routes = self.state
+            .filterMap { $0.route } // State to Route conversion
+    }
+
+    // User Action from UI
+    func send(action: Action) {
+        eventPipe.observer(value: .ui(action))
+    }
+
+    // MARK: - Reducer
+
+    static func reduce(_ state: State, _ event: Event) -> State {
+        switch (oldState, event) {
+        case ...:
+            return newState
+        }
+    }
+
+    // MARK: - Feedback
+
+    static func whenSubmitting(...) {
+        return Feedback(
+            predicate: { $0.status.isSubmitting },
+            effects: { state -> SignalProducer<Event, NoError> in
+                return self.validate(state: state)
+                    .flatMap(.merge) { data in
+                        self.fetchNextScreen(data)
+                    }
+                    .materializeResults()
+                    .map(Event.didFinishSubmit)
+            }
+        )
+    }
+}
+```
+
+- Input: User actions from `Renderer` will be delivered to `func send(action: Action)`, and forwarded as ReactiveFeedback's input.
```

Comment:
Should we also mention other sources of input? For example API responses, permission prompt responses, etc.

Reply:
This PR focuses on `Describe how VMs, VC, Renderer (Bento), FC and Builders interact` , so I think other external effects don't have to be in this document for simplicity.